src/gui/plugins/visualize_lidar/VisualizeLidar.cc

```diff
@@ -0,0 +1,524 @@
+        }
+        if (success)
+        {
+          this->dataPtr->lidarEntity = parent;
+          this->dataPtr->lidarEntityDirty = false;
+        }
+      }
+    }
+  }
+
+  // Only update lidarPose if the lidarEntity exists and the lidar is initialized
+  // and the sensor message is yet to arrive.
+  //
+  // If we update the worldpose on the physics thread **after** the sensor
+  // data arrives, the visual is offset from the obstacle if the sensor is
+  // moving fast.
+  if (!this->dataPtr->lidarEntityDirty && this->dataPtr->initialized &&
+                                          !this->dataPtr->visualDirty)
+  {
+    this->dataPtr->lidarPose = worldPose(this->dataPtr->lidarEntity, _ecm);
+  }
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::UpdateType(int _type)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+  switch (_type)
+  {
+    case 0:
+      this->dataPtr->visualType = rendering::LidarVisualType::LVT_NONE;
+      break;
+
+    case 1:
+      this->dataPtr->visualType = rendering::LidarVisualType::LVT_RAY_LINES;
+      break;
+
+    case 2:
+      this->dataPtr->visualType = rendering::LidarVisualType::LVT_POINTS;
+      break;
+
+    case 3:
+      this->dataPtr->visualType =
+                            rendering::LidarVisualType::LVT_TRIANGLE_STRIPS;
+      break;
+
+    default:
+      this->dataPtr->visualType =
+                            rendering::LidarVisualType::LVT_TRIANGLE_STRIPS;
+      break;
+  }
+  this->dataPtr->lidar->SetType(this->dataPtr->visualType);
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::OnTopic(const QString &_topicName)
+{
+  std::lock_guard<std::mutex>(this->dataPtr->serviceMutex);
+  if (!this->dataPtr->node.Unsubscribe(this->dataPtr->topicName))
+  {
+    ignerr << "Unable to unsubscribe from topic ["
+           << this->dataPtr->topicName <<"]" <<std::endl; 
+  }
+  this->dataPtr->topicName = _topicName.toStdString();
+
+  // Reset visualization
+  this->dataPtr->resetVisual = true;
+
+  // Create new subscription
+  if (!this->dataPtr->node.Subscribe(this->dataPtr->topicName,
+                            &VisualizeLidar::OnScan, this))
+  {
+    ignerr << "Unable to subscribe to topic ["
+           << this->dataPtr->topicName << "]\n";
+    return;
+  }
+  this->dataPtr->visualDirty = false;
+  ignmsg << "Subscribed to " << this->dataPtr->topicName << std::endl;
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::UpdateNonHitting(bool _value)
+{
+  std::lock_guard<std::mutex>(this->dataPtr->serviceMutex);
+  this->dataPtr->lidar->SetDisplayNonHitting(_value);
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::DisplayVisual(bool _value)
+{
+  std::lock_guard<std::mutex>(this->dataPtr->serviceMutex);
+  if (!_value)
+  {
+    ignmsg << "Lidar Visual Display OFF." << std::endl;
+    this->dataPtr->lidar->SetType(rendering::LidarVisualType::LVT_NONE);
```

Comment:
you should be able to call `this->dataPtr->lidar->SetVisible(false)` to make it invisible, which is a faster operation than clearing all the lidar data. Same for re-enabling the lidar visual display below